src/lib/message/L2ToL1Message.ts

```diff
@@ -312,6 +315,71 @@ export class L2ToL1MessageReader extends L2ToL1Message {
       await this.waitUntilOutboxEntryCreated(retryDelay)
     }
   }
+
+  // TODO: create version that queries multiple L2 to L1 txs, so a single multicall can make all requests
+  public async getExpectedNodeDeadline(
+    l2Provider: Provider
+  ): Promise<BigNumber> {
+    const outbox = Outbox__factory.connect(this.outboxAddress, this.l1Provider)
+
+    const network = l2Networks[(await l2Provider.getNetwork()).chainId]
+    const maxTime = BigNumber.from(network.confirmPeriodBlocks).mul(
+      BigNumber.from(l1Networks[network.partnerChainID].blockTime)
+    )
+
+    // TODO: use IRollupUser interface instead
+    const rollup = RollupUserFacet__factory.connect(
+      network.ethBridge.rollup,
+      this.l1Provider
+    )
+
+    try {
+      const proof = await this.tryGetProof(l2Provider)
+      // here we assume the L2 to L1 tx is actually valid, so the user needs to wait the max time.
+      // we could check that on the constructor that the L2 to L1 msg is valid
+      if (proof === null) return maxTime
+      // we can't check if the L2 to L1 tx isSpent on the outbox, so we instead try executing it
+      if (await this.hasExecuted(proof)) return BigNumber.from(0)
+      // 60seconds * 60 minutes * 24hrs * 8 days
+      const EIGHT_DAYS_IN_BLOCKS =
+        (60 * 60 * 24 * 8) / l1Networks[network.partnerChainID].blockTime
+      const latestBlock = await this.l1Provider.getBlockNumber()
+
+      const eventFetcher = new EventFetcher(this.l1Provider)
+
+      const events = (
+        await eventFetcher.getEvents(
+          network.ethBridge.rollup,
+          RollupUserFacet__factory,
+          t => t.filters.NodeCreated(),
+          {
+            fromBlock: latestBlock - EIGHT_DAYS_IN_BLOCKS,
+            toBlock: latestBlock,
```

Comment:
could also separate this into a 10 different queries, each being 4000 block range. dispatch in parallel etc